Vector.c

```diff
@@ -242,6 +261,56 @@ Object* Vector_remove(Vector* this, int idx) {
    }
 }
 
+Object* Vector_softRemove(Vector* this, int idx) {
+   assert(idx >= 0 && idx < this->items);
+
+   Object* removed = this->array[idx];
+   this->array[idx] = NULL;
+
+   this->dirty_count++;
+   if (this->dirty_index < 0 || idx < this->dirty_index) {
+      this->dirty_index = idx;
+   }
+
+   if (this->owner) {
+      Object_delete(removed);
+      return NULL;
+   }
+
+   return removed;
+}
+
+void Vector_compact(Vector* this) {
+   if (!Vector_isDirty(this)) {
+      return;
+   }
+
+   const int size = this->items;
+   assert(0 <= this->dirty_index && this->dirty_index < size);
+   assert(this->array[this->dirty_index] == NULL);
+
+   int idx = this->dirty_index;
+
+   /* one deletion: use memmove, which should be faster */
+   if (this->dirty_count == 1) {
+      memmove(&this->array[idx], &this->array[idx + 1], (this->items - idx) * sizeof(this->array[0]));
+   } else {
+      /* multiple deletions */
+      for (int i = this->dirty_index + 1; i < size; i++) {
+         if (this->array[i]) {
+            this->array[idx++] = this->array[i];
+         }
+      }
+   }
+
+   this->items -= this->dirty_count;
+   this->dirty_index = -1;
+   this->dirty_count = 0;
+
+   assert(Vector_count(this) == (unsigned int)this->items);
```

Comment:
This can be dropped, because it is enforced implicitly by the `Vector_isConsistent` check below.

Reply:
Fixed